File: test/node/helper/merchantBniVaTopUp.ts
import Dana from 'dana-node';
import {
  QueryAssetCardListRequestAssetTypeListEnum,
  QueryAssetCardListRequestEnableOnlyEnum,
} from 'dana-node/merchant_management/v1';
import dotenv from 'dotenv';
import { hashData } from './bniHashUtil';

dotenv.config();

const DANA_SANDBOX_BASE_URL = 'https://api.sandbox.dana.id';
const BNI_VA_TOP_UP_PATH = '/ifcsupergw/bni/topup/merchant/request.htm';

const DEFAULT_BNI_CLIENT_ID = '910';
const DEFAULT_BNI_SECRET_KEY = '9546d5f69af2ed3bc603834446985628';
const BNI_INST_ID = 'BNIC1ID';
const MERCHANT_DEPOSIT_ACCOUNT_TYPE = 'MERCHANT_DEPOSIT_ACCOUNT';
const MERCHANT_QUERY_LOGIN_TYPE_ROLE = 'ROLE';
const MERCHANT_DEPOSIT_TOP_UP_THRESHOLD = 1_000_000;

let done = false;
let failure: Error | null = null;
let ensurePromise: Promise<void> | null = null;
let danaClient: Dana | null = null;

function getDanaClient(): Dana {
  if (!danaClient) {
    danaClient = new Dana({
      partnerId: process.env.X_PARTNER_ID || '',
      privateKey: process.env.PRIVATE_KEY || '',
      origin: process.env.ORIGIN || '',
      env: process.env.ENV || 'sandbox',
      clientSecret: process.env.CLIENT_SECRET || '',
    });
  }
  return danaClient;
}

export async function ensureMerchantBniVaTopUp(): Promise<void> {
  if (done) {
    if (failure) {
      throw failure;
    }
    return;
  }
  if (!ensurePromise) {
    ensurePromise = (async () => {
      try {
        await bniVaTopUpMerchantOnce();
      } catch (err) {
        failure = err instanceof Error ? err : new Error(String(err));
        throw failure;
      } finally {
        done = true;
      }
    })();
  }
  await ensurePromise;
}

async function bniVaTopUpMerchantOnce(): Promise<void> {
  const depositBalance = await queryMerchantDepositTotalAmount();
  if (depositBalance >= MERCHANT_DEPOSIT_TOP_UP_THRESHOLD) {
    console.log(
      `Skipping BNI VA top-up: merchant deposit balance=${depositBalance} >= threshold=${MERCHANT_DEPOSIT_TOP_UP_THRESHOLD}`,
    );
    return;
  }
  console.log(
    `Merchant deposit balance=${depositBalance} < threshold=${MERCHANT_DEPOSIT_TOP_UP_THRESHOLD}; proceeding with BNI VA top-up`,
  );

  const virtualAccount = await queryBniMerchantVirtualAccount();
  await postBniVaTopUpMerchant(virtualAccount);
}

async function queryMerchantDepositTotalAmount(): Promise<number> {
  const merchantId = process.env.MERCHANT_ID || '';
  if (!merchantId) {
    throw new Error('MERCHANT_ID is required to query merchant info');
  }

  const response = await getDanaClient().merchantManagementApi.queryMerchantInfo({
    roleId: merchantId,
    loginType: MERCHANT_QUERY_LOGIN_TYPE_ROLE,
    isQueryAccount: true,
  });

  const resultInfo = response?.response?.body?.resultInfo;
  if (resultInfo?.resultStatus !== 'S') {
    throw new Error(`queryMerchantInfo failed: ${JSON.stringify(resultInfo)}`);
  }

  const accounts = response?.response?.body?.merchantInformation?.accounts || [];
  for (const account of accounts) {
    if (account?.accountType === MERCHANT_DEPOSIT_ACCOUNT_TYPE) {
      return parseAccountMappedTotalAmount(account as Record<string, unknown>);
    }
  }

  throw new Error(`queryMerchantInfo: ${MERCHANT_DEPOSIT_ACCOUNT_TYPE} account not found`);
}

function parseAccountMappedTotalAmount(account: Record<string, unknown>): number {
  const mappedTotal = account.mappedTotalAmount;
  if (mappedTotal && typeof mappedTotal === 'object' && mappedTotal !== null && 'amount' in mappedTotal) {
    return parseAmountValue((mappedTotal as { amount: unknown }).amount);
  }

  const totalAmount = account.totalAmount;
  if (typeof totalAmount === 'string' && totalAmount) {
    const parsed = JSON.parse(totalAmount) as { amount?: unknown };
    return parseAmountValue(parsed.amount);
  }

  throw new Error('deposit account amount missing in queryMerchantInfo response');
}

function parseAmountValue(value: unknown): number {
  if (typeof value === 'string') {
    return Number.parseInt(value, 10);
  }
  if (typeof value === 'number') {
    return Math.trunc(value);
  }
  throw new Error(`unsupported amount format: ${String(value)}`);
}

async function queryBniMerchantVirtualAccount(): Promise<string> {
  const memberId = process.env.MERCHANT_ID || '';
  if (!memberId) {
    throw new Error('MERCHANT_ID is required to query merchant BNI VA');
  }

  const response = await getDanaClient().merchantManagementApi.queryAssetCardList({
    memberId,
    enableOnly: QueryAssetCardListRequestEnableOnlyEnum.True,
    assetTypeList: [QueryAssetCardListRequestAssetTypeListEnum.VaAccount],
  });

  const resultInfo = response?.response?.body?.resultInfo;
  if (resultInfo?.resultStatus !== 'S') {
    throw new Error(`queryAssetCardList failed: ${JSON.stringify(resultInfo)}`);
  }

  for (const card of response?.response?.body?.assetCardList || []) {
    if (card?.assetType === 'VA_ACCOUNT' && card?.instId === BNI_INST_ID) {
      const cardIndexNo = card?.cardIndexNo || '';
      if (cardIndexNo) {
        return cardIndexNo;
      }
    }
  }

  throw new Error(`BNI VA card not found in assetCardList (assetType=VA_ACCOUNT instId=${BNI_INST_ID})`);
}

async function postBniVaTopUpMerchant(virtualAccount: string): Promise<void> {
  const { trxId, datetimePayment, datetimePaymentIso } = jakartaTimestamps();

  const integrationBody = {
    trx_amount: '1000',
    trx_id: trxId,
    virtual_account: virtualAccount,
    customer_name: 'rudy',
    payment_amount: '1000000000',
    cumulative_payment_amount: '1000',
    payment_ntb: '233171',
    datetime_payment: datetimePayment,
    datetime_payment_iso8601: datetimePaymentIso,
  };

  const clientId = process.env.BNI_VA_TOP_UP_CLIENT_ID || DEFAULT_BNI_CLIENT_ID;
  const secretKey = process.env.BNI_VA_TOP_UP_SECRET_KEY || DEFAULT_BNI_SECRET_KEY;
  const integrationJson = JSON.stringify(integrationBody);
  const data = hashData(integrationJson, clientId, secretKey);

  const payload = JSON.stringify({ client_id: clientId, data });
  await httpPostJson(DANA_SANDBOX_BASE_URL + BNI_VA_TOP_UP_PATH, payload, true);
}

async function httpPostJson(url: string, body: string, checkBniStatus: boolean): Promise<string> {
  const res = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body,
  });

  const responseText = await res.text();
  if (!res.ok) {
    throw new Error(`HTTP ${res.status}: ${responseText}`);
  }

  if (checkBniStatus) {
    try {
      const parsed = JSON.parse(responseText);
      if (parsed?.status && parsed.status !== '' && parsed.status !== '000') {
        throw new Error(`BNI VA top-up rejected: ${responseText}`);
      }
    } catch (err) {
      if (err instanceof SyntaxError) {
        return responseText;
      }
      throw err;
    }
  }

  return responseText;
}

function jakartaTimestamps(): { trxId: string; datetimePayment: string; datetimePaymentIso: string } {
  const formatter = new Intl.DateTimeFormat('en-GB', {
    timeZone: 'Asia/Jakarta',
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    hour12: false,
  });

  const parts = formatter.formatToParts(new Date());
  const get = (type: Intl.DateTimeFormatPartTypes) =>
    parts.find((p) => p.type === type)?.value || '00';

  const year = get('year');
  const month = get('month');
  const day = get('day');
  const hour = get('hour');
  const minute = get('minute');
  const second = get('second');

  return {
    trxId: `${year}${month}${day}${hour}${minute}${second}`,
    datetimePayment: `${year}-${month}-${day} ${hour}:${minute}:${second}`,
    datetimePaymentIso: `${year}-${month}-${day}T${hour}:${minute}:${second}+07:00`,
  };
}

File: test/go/helper/merchant_bni_va_topup.go
package helper

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/dana-id/dana-go/v2/merchant_management/v1"
)

const (
	danaSandboxBaseURL          = "https://api.sandbox.dana.id"
	bniVATopUpMerchantPath      = "/ifcsupergw/bni/topup/merchant/request.htm"
	defaultBNIVATopUpClientID   = "910"
	defaultBNIVATopUpKey  = "9546d5f69af2ed3bc603834446985628"
	merchantVAAccountAssetType      = "VA_ACCOUNT"
	merchantBNIVAInstID             = "BNIC1ID"
	merchantDepositAccountType      = "MERCHANT_DEPOSIT_ACCOUNT"
	merchantQueryLoginTypeRole      = "ROLE"
	merchantDepositTopUpThreshold   = int64(1_000_000)
)

var (
	merchantTopUpOnce sync.Once
	merchantTopUpErr  error
)

// EnsureMerchantBNIVATopUp tops up merchant deposit balance via BNI VA before disbursement tests.
func EnsureMerchantBNIVATopUp() error {
	merchantTopUpOnce.Do(func() {
		merchantTopUpErr = bniVATopUpMerchantOnce()
	})
	return merchantTopUpErr
}

func bniVATopUpMerchantOnce() error {
	depositBalance, err := queryMerchantDepositTotalAmount()
	if err != nil {
		return err
	}
	if depositBalance >= merchantDepositTopUpThreshold {
		fmt.Printf(
			"Skipping BNI VA top-up: merchant deposit balance=%d >= threshold=%d\n",
			depositBalance, merchantDepositTopUpThreshold,
		)
		return nil
	}
	fmt.Printf(
		"Merchant deposit balance=%d < threshold=%d; proceeding with BNI VA top-up\n",
		depositBalance, merchantDepositTopUpThreshold,
	)

	virtualAccount, err := queryBNIMerchantVirtualAccount()
	if err != nil {
		return err
	}

	now := time.Now()
	trxID := now.Format("20060102150405")
	datetimePayment := now.Format("2006-01-02 15:04:05")
	datetimePaymentISO := now.Format("2006-01-02T15:04:05-07:00")

	integrationBody := map[string]string{
		"trx_amount":                  "1000",
		"trx_id":                      trxID,
		"virtual_account":             virtualAccount,
		"customer_name":               "rudy",
		"payment_amount":              "1000000",
		"cumulative_payment_amount":   "1000",
		"payment_ntb":                 "233171",
		"datetime_payment":            datetimePayment,
		"datetime_payment_iso8601":    datetimePaymentISO,
	}

	return postBNIVATopUpMerchant(integrationBody)
}

func queryMerchantDepositTotalAmount() (int64, error) {
	merchantID := os.Getenv("MERCHANT_ID")
	if merchantID == "" {
		return 0, fmt.Errorf("MERCHANT_ID is required to query merchant info")
	}

	isQueryAccount := true
	req := merchant_management.NewQueryMerchantInfoRequest(merchantID, merchantQueryLoginTypeRole)
	req.SetIsQueryAccount(isQueryAccount)

	ctx := context.Background()
	resp, httpRes, err := ApiClient.MerchantManagementAPI.
		QueryMerchantInfo(ctx).
		QueryMerchantInfoRequest(*req).
		Execute()
	if err != nil {
		return 0, fmt.Errorf("queryMerchantInfo: %w", err)
	}
	if httpRes != nil {
		defer httpRes.Body.Close()
	}
	if resp == nil {
		return 0, fmt.Errorf("queryMerchantInfo: empty response")
	}

	responseObj := resp.GetResponse()
	body := responseObj.GetBody()
	resultInfo := body.GetResultInfo()
	if resultInfo.GetResultStatus() != "S" {
		return 0, fmt.Errorf("queryMerchantInfo failed: status=%s code=%s msg=%s",
			resultInfo.GetResultStatus(), resultInfo.GetResultCodeId(), resultInfo.GetResultMsg())
	}

	merchantInfo, ok := body.GetMerchantInformationOk()
	if !ok || merchantInfo == nil {
		return 0, fmt.Errorf("queryMerchantInfo: merchantInformation missing")
	}

	for _, account := range merchantInfo.GetAccounts() {
		if account.GetAccountType() != merchantDepositAccountType {
			continue
		}
		amount, err := parseAccountMappedTotalAmount(account)
		if err != nil {
			return 0, fmt.Errorf("parse deposit account amount: %w", err)
		}
		return amount, nil
	}

	return 0, fmt.Errorf("queryMerchantInfo: %s account not found", merchantDepositAccountType)
}

func parseAccountMappedTotalAmount(account merchant_management.MerchantAccountInfo) (int64, error) {
	// Prefer mappedTotalAmount when present (API may return it even if not in SDK struct).
	if raw, err := json.Marshal(account); err == nil {
		var extra struct {
			MappedTotalAmount *struct {
				Amount json.RawMessage `json:"amount"`
			} `json:"mappedTotalAmount"`
		}
		if err := json.Unmarshal(raw, &extra); err == nil && extra.MappedTotalAmount != nil {
			return parseAmountValue(extra.MappedTotalAmount.Amount)
		}
	}

	return parseAccountTotalAmountJSON(account.GetTotalAmount())
}

func parseAccountTotalAmountJSON(totalAmountJSON string) (int64, error) {
	if totalAmountJSON == "" {
		return 0, fmt.Errorf("empty totalAmount")
	}
	var parsed struct {
		Amount json.RawMessage `json:"amount"`
	}
	if err := json.Unmarshal([]byte(totalAmountJSON), &parsed); err != nil {
		return 0, err
	}
	return parseAmountValue(parsed.Amount)
}

func parseAmountValue(raw json.RawMessage) (int64, error) {
	if len(raw) == 0 {
		return 0, fmt.Errorf("empty amount")
	}

	var asString string
	if err := json.Unmarshal(raw, &asString); err == nil {
		return strconv.ParseInt(asString, 10, 64)
	}

	var asFloat float64
	if err := json.Unmarshal(raw, &asFloat); err == nil {
		return int64(asFloat), nil
	}

	var asInt int64
	if err := json.Unmarshal(raw, &asInt); err == nil {
		return asInt, nil
	}

	return 0, fmt.Errorf("unsupported amount format: %s", string(raw))
}

func queryBNIMerchantVirtualAccount() (string, error) {
	memberID := os.Getenv("MERCHANT_ID")
	if memberID == "" {
		return "", fmt.Errorf("MERCHANT_ID is required to query merchant BNI VA")
	}

	enableOnly := "true"
	req := merchant_management.NewQueryAssetCardListRequest(memberID)
	req.SetEnableOnly(enableOnly)
	req.SetAssetTypeList([]string{merchantVAAccountAssetType})

	ctx := context.Background()
	resp, httpRes, err := ApiClient.MerchantManagementAPI.
		QueryAssetCardList(ctx).
		QueryAssetCardListRequest(*req).
		Execute()
	if err != nil {
		return "", fmt.Errorf("queryAssetCardList: %w", err)
	}
	if httpRes != nil {
		defer httpRes.Body.Close()
	}

	if resp == nil {
		return "", fmt.Errorf("queryAssetCardList: empty response")
	}

	responseObj := resp.GetResponse()
	body := responseObj.GetBody()
	resultInfo := body.GetResultInfo()
	if resultInfo.GetResultStatus() != "S" {
		return "", fmt.Errorf("queryAssetCardList failed: status=%s code=%s msg=%s",
			resultInfo.GetResultStatus(), resultInfo.GetResultCodeId(), resultInfo.GetResultMsg())
	}

	cards := body.GetAssetCardList()
	for _, card := range cards {
		if card.AssetType == merchantVAAccountAssetType && card.InstId == merchantBNIVAInstID {
			if card.CardIndexNo != "" {
				return card.CardIndexNo, nil
			}
		}
	}

	return "", fmt.Errorf("BNI VA card not found in assetCardList (assetType=%s instId=%s)", merchantVAAccountAssetType, merchantBNIVAInstID)
}

func postBNIVATopUpMerchant(integrationBody map[string]string) error {
	clientID := envOrDefault("BNI_VA_TOP_UP_CLIENT_ID", defaultBNIVATopUpClientID)
	secretKey := envOrDefault("BNI_VA_TOP_UP_SECRET_KEY", defaultBNIVATopUpKey)

	integrationJSON, err := json.Marshal(integrationBody)
	if err != nil {
		return err
	}

	data, err := hashBNIData(string(integrationJSON), clientID, secretKey)
	if err != nil {
		return err
	}

	payload := map[string]string{
		"client_id": clientID,
		"data":      data,
	}
	body, err := json.Marshal(payload)
	if err != nil {
		return err
	}

	url := danaSandboxBaseURL + bniVATopUpMerchantPath
	httpReq, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return err
	}
	httpReq.Header.Set("Content-Type", "application/json")

	client := &http.Client{Timeout: 30 * time.Second}
	httpRes, err := client.Do(httpReq)
	if err != nil {
		return fmt.Errorf("BNI VA top-up request: %w", err)
	}
	defer httpRes.Body.Close()

	respBody, _ := io.ReadAll(httpRes.Body)
	if httpRes.StatusCode < 200 || httpRes.StatusCode >= 300 {
		return fmt.Errorf("BNI VA top-up failed: status=%d body=%s", httpRes.StatusCode, string(respBody))
	}

	var parsed map[string]interface{}
	if err := json.Unmarshal(respBody, &parsed); err == nil {
		if status, ok := parsed["status"].(string); ok && status != "" && status != "000" {
			return fmt.Errorf("BNI VA top-up rejected: status=%s body=%s", status, string(respBody))
		}
	}

	return nil
}

func hashBNIData(jsonData, clientID, secretKey string) (string, error) {
	timeStr := fmt.Sprintf("%d", time.Now().UnixMilli())
	if len(timeStr) > 10 {
		timeStr = timeStr[:10]
	}
	runes := []rune(timeStr)
	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	payload := string(runes) + "." + jsonData
	return doubleEncrypt(payload, clientID, secretKey), nil
}

func encrypt(data []byte, key string) []byte {
	result := make([]byte, len(data))
	keyLen := len(key)
	for i := range data {
		keyChar := key[(i+keyLen-1)%keyLen]
		result[i] = byte((int(data[i]) + int(keyChar)) % 128)
	}
	return result
}

func doubleEncrypt(input, clientID, secretKey string) string {
	encrypted := encrypt([]byte(input), clientID)
	encrypted = encrypt(encrypted, secretKey)
	encoded := base64.StdEncoding.EncodeToString(encrypted)
	encoded = strings.TrimRight(encoded, "=")
	encoded = strings.ReplaceAll(encoded, "+", "-")
	encoded = strings.ReplaceAll(encoded, "/", "_")
	return encoded
}

func envOrDefault(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}
